src/build_context.rs

```diff
@@ -1,133 +1,104 @@
+    definitions: Vec<Option<T>>,
 }
 
 impl<T: Clone + std::fmt::Debug> BuildContext<T> {
-    pub fn new(schema: &PyAny) -> PyResult<Self> {
-        let mut used_refs = AHashSet::new();
-        extract_used_refs(schema, &mut used_refs)?;
+    pub fn new(_schema: &PyAny) -> PyResult<Self> {
         Ok(Self {
-            used_refs,
-            slots: Vec::new(),
-            reusable: AHashMap::new(),
+            refs: AHashMap::new(),
+            definitions: Vec::new(),
         })
     }
 
-    pub fn for_self_schema() -> Self {
-        let mut used_refs = AHashSet::with_capacity(3);
-        // NOTE: we don't call `extract_used_refs` for performance reasons, if more recursive references
-        // are used, they would need to be manually added here.
-        // we use `2` as count to avoid `find_slot` pulling the validator out of slots and returning it directly
-        used_refs.insert("root-schema".to_string());
-        used_refs.insert("ser-schema".to_string());
-        used_refs.insert("inc-ex-type".to_string());
-        Self {
-            used_refs,
-            slots: Vec::new(),
-            reusable: AHashMap::new(),
-        }
-    }
-
-    /// Check whether a ref is already in `reusable` or `slots`, we shouldn't allow repeated refs
-    pub fn ref_already_used(&self, ref_: &str) -> bool {
-        self.reusable.contains_key(ref_) || self.slots.iter().any(|slot| slot.slot_ref == ref_)
-    }
-
-    /// check if a ref is used elsewhere in the schema
-    pub fn ref_used(&self, ref_: &str) -> bool {
-        self.used_refs.contains(ref_)
-    }
-
-    /// check if a ref is used within a given schema
-    pub fn ref_used_within(&self, schema_dict: &PyAny, ref_: &str) -> PyResult<bool> {
-        check_ref_used(schema_dict, ref_)
-    }
-
-    /// add a validator/serializer to `reusable` so it can be cloned and used again elsewhere
-    pub fn store_reusable(&mut self, ref_: String, val_ser: T) {
-        self.reusable.insert(ref_, val_ser);
+    fn get_ref_from_id(&self, ref_id: usize) -> &String {
+        // We can unwrap here because nothing should ever have a ref_id unless they've inserted it.
+        self.refs
+            .iter()
+            .find_map(|(key, &val)| if val == ref_id { Some(key) } else { None })
+            .unwrap()
     }
 
-    /// First of two part process to add a new validator/serializer slot, we add the `slot_ref` to the array,
-    /// but not the actual `validator`/`serializer`, we can't add that until it's build.
-    /// But we need the `id` to build it, hence this two-step process.
-    pub fn prepare_slot(&mut self, slot_ref: String) -> PyResult<usize> {
-        let id = self.slots.len();
-        let slot = Slot {
-            slot_ref,
-            op_val_ser: None,
-        };
-        self.slots.push(slot);
-        Ok(id)
+    pub fn get_ref_id(&mut self, reference: String) -> usize {
```

Comment:
can we not use `&str` here?

Reply:
The entry API requires a String, you get to choose between two hashmap lookups or a string copy

Reply:
ok.